Allow MFEM coefficients to take numeric values

cmacmackin opened the pull request:
## Reason
MOOSE functors will accept numerical values, in addition to names of functions etc. Previously MFEM coefficients (which are used analogously to functors) did not do this. Allowing them to accept numeric values also means we no longer need the constant versions of materials and boundary conditions, reducing the amount of code that needs to be maintained.

## Design
When the CoefficientManager class tries to look up a coefficient (e.g., in `getXXXCoefficient` or when declaring an alias for a coefficient) which it finds has not been declared, it will now attempt to interpret the name of the coefficient as a numeric value. For scalar coefficients, it will attempt to read it as a single float. For vectors it will try to parse the name as a space-separated list of floats. If this succeeds then it will create and return a Constant(Vector)Coefficient with the corresponding value. This behaviour is loosely modeled after how functors are looked up in the main part of MOOSE.

Currently there is no support for using numeric values with matrix coefficients, as it was not clear what would be the appropriate syntax for that. However, matrix coefficients are not actually used anywhere yet anyway.

## Impact
This PR removes the following components, as their functionality is now redundant. All tests have been updated to reflect these changes.

- MFEMScalarDirichletBC (use MFEMScalarFunctorDirichletBC)
- MFEMVectorBoundaryIntegratedBC (use MFEMVectorFunctorBoundaryIntegratedBC)
- MFEMVectorDirichletBC (use MFEMVectorFunctorDirichletBC)
- MFEMVectorDirichletBCBase (use MFEMVectorFunctorDirichletBCBase)
- MFEMVectorNormalDirichletBC (use MFEMVectorFunctorNormalDirichletBC)
- MFEMVectorNormalIntegratedBC (use MFEMVectorFunctorNormalIntegratedBC)
- MFEMGenericConstantFunctorMaterial (use MFEMGenericFunctorMaterial)
- MFEMGenericConstantVectorFunctorMaterial (use MFEMGenericFunctorVectorMaterial)

One tricky aspect of these changes were how to allow MFEMGenericFunctorVectorMaterial to take a mix of function names (etc.) and numeric vector values for `prop_values`. The numeric values will typically require multiple space-separated components to describe the vector, e.g.:
```
prop_names = mag_field gravity
prop_values = func1 0 0 9.8
```
The HIT parser will interpret prop_values as `std::vector<MFEMVectorCoefficientName>({"func1", "0", "0", "9.8"})` and there is no general way to associate each item with the appropriate `prop_name`. (E.g., Consider the case of multiple numeric properties  being defined, with vectors of different dimensions). The solution adopted was to add a preprocessing step for this parameter, that will group items enclosed in curly braces. So the input 
```
prop_names = mag_field gravity
prop_values = func1 {0 0 9.8}
```
will be processed into `std::vector<MFEMVectorCoefficientName>({"func1", "0 0 9.8"})`. This appears to work well, but does have the disadvantage of introducing new syntactic conventions.

cmacmackin committed:
Added ability to create coefficients from numbers

cmacmackin committed:
Update to docs to explain parsing of constant coefficients

cmacmackin committed:
Removed redundant constant versions of BCs and (scalar) materials

The unit tests have been updated to reflect this change. The
integration tests still need to be changed, however. I'm not 100% sure
on how vectors of numbers can be grouped together for input to the
MFEMGenericFunctorVectorMaterial class, so I haven't gotten rid of the
constant version of that yet.

cmacmackin committed:
Refactored MFEM integration tests for new functor classes

cmacmackin committed:
Adapted MFEMGenericFunctorVectorMaterial to take numeric literals

cmacmackin commented:
@alexanderianblair @loganharbour @lindsayad You would probably all be good people to take a look at this.

One question I have is whether we should consider changing the names of things like `MFEMScalarFunctorDirichletBC`, now that we no longer have a non-functor version. Should we rename it to `MFEMScalarDirichletBC`? That would be a bit shorter and less wordy. The main downside is that we previously had a different object with that name and this might cause confusion and more subtle breaking changes for users.

cmacmackin committed:
Fixed typo in PR idaholab/moose#30622

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2910397/), step Clang format on f598c29 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30622/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30622/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7cc8bd7530c04b73ef16019251092c5a42403fdc`

cmacmackin committed:
Applied styles updates from CI

cmacmackin committed:
Ensure all requirements included directly in MooseStringUtils.h

cmacmackin committed:
Fixed broken references in docs

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2917174/), step Docs: sync website on 7a42301 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30622/site)

This comment will be updated on new commits.

alexanderianblair commented:
> Can we at least come up with a reasonable naming scheme (even though we don't know if it'll fit the design yet) that would align with this change?

The simplest here would be `MFEMADXXXKernel` / `MFEMADXXXBC`, since that'll be our method of AD support, and would avoid a name clash with either of these options.

However, for context - the dFEM-like method of specifying kernel contributions appears to be much closer in philosophy to how a regular MOOSE kernel is written (with the additional ability to support the full range of GPU assembly levels), so I anticipate/hope that we'll eventually use those as our preferred kernel/BCs over our current `mfem::XXXFormIntegrator`-based ones, at which point the simpler `MFEMXXXKernel` designation would better describe them instead, and the current batch migrate to using the `MFEMXXXFunctorKernel` naming convention. 

So pros/cons to both options - using `MFEMXXXKernel` for the current integrator-type batch and `MFEMADXXXKernel` for the dFEM ones is clearer for users based on the current state IMO, but using `MFEMXXXFunctorKernel` for the current integrator-type batch and `MFEMXXXKernel` for future dFEM ones aligns closer to the design philosophy of the analagous MOOSE objects, if things pan out as expected.

lindsayad commented:
> However, for context - the dFEM-like method of specifying kernel contributions appears to be much closer in philosophy to how a regular MOOSE kernel is written

Can you elaborate on this? I believe when we've talked in the past you've said that MOOSE is at too high of an assembly level (element/E level in libCEED terms) vs. being at the quadrature-point/Q level. Is this what you're talking about?

cmacmackin commented:
There is still one failing CI job. Looking at the results, I'm not sure why that would happen as it doesn't involve any of the code I've changed. The particular test that failed also appears to have passed in other jobs. Should I just try marking that job as invalid and running again?

loganharbour commented:
> There is still one failing CI job. Looking at the results, I'm not sure why that would happen as it doesn't involve any of the code I've changed. The particular test that failed also appears to have passed in other jobs. Should I just try marking that job as invalid and running again?

The failure in https://civet.inl.gov/job/2917216/ is unrelated. Fix is in next and it can be ignored.

cmacmackin commented:
@lindsayad @loganharbour What are your feelings on the renaming issue? It would be good to get this decided so we can merge the PR.

cmacmackin committed:
Renamed MFEM functor BCs

cmacmackin committed:
Renamed MFEM BCs in tests and docs

cmacmackin committed:
Applied clang-format

cmacmackin committed:
Removed MFEMGenericConstantFunctorMaterial from tests

cmacmackin committed:
Removed old-style constant BCs

lindsayad commented:
We have a PR up to fix that test at https://github.com/idaholab/moose/pull/30748

lindsayad reviewed (changes requested):
37/78 through

cmacmackin committed:
Fix oddities in vector Dirichlet BC

cmacmackin commented:
I've implemented the requested changes. Note that these issues predate this PR; they just showed up in the diff because git got confused about the fact I renamed a file.

cmacmackin committed:
Added listing to show use of constant value for BC

lindsayad commented on framework/include/mfem/functormaterials/MFEMGenericFunctorVectorMaterial.h:
```diff
@@ -19,7 +19,7 @@ class MFEMGenericFunctorVectorMaterial : public MFEMFunctorMaterial
 protected:
   const std::vector<std::string> & _prop_names;
-  const std::vector<MFEMVectorCoefficientName> & _prop_values;
+  const std::vector<MFEMVectorCoefficientName> _prop_values;
```
I'd revert this change. No reason to store this twice, once in the input parameters warehouse and once here

lindsayad replied:
Oops I was looking at the wrong local repo. I see that this does indeed have to be its own storage

lindsayad commented on framework/src/mfem/bcs/MFEMVectorTangentialDirichletBC.C:
```diff
@@ -4,15 +4,6 @@
 
 registerMooseObject("MooseApp", MFEMVectorTangentialDirichletBC);
 
-InputParameters
-MFEMVectorTangentialDirichletBC::validParams()
-{
-  InputParameters params = MFEMVectorDirichletBCBase::validParams();
-  params.addClassDescription(
-      "Applies a Dirichlet condition to the tangential components of a vector variable.");
-  return params;
-}
```
This should also stay

lindsayad commented on framework/src/mfem/functormaterials/MFEMGenericFunctorVectorMaterial.C:
```diff
@@ -5,6 +5,59 @@
+    if (in_literal)
+    {
+      if (item.front() == '{')
+      {
+        mooseError("Nested numeric vector values are not permitted in "
+                   "MFEMGenericFunctorVectoMaterial prop_values.");
+      }
```
```suggestion
        mooseError("Nested numeric vector values are not permitted in "
                   "MFEMGenericFunctorVectoMaterial prop_values.");
```

cmacmackin replied:
Actually, seeing as you've now committed these in #30803 we might as well just merge that.

lindsayad commented on framework/src/mfem/functormaterials/MFEMGenericFunctorVectorMaterial.C:
```diff
@@ -5,6 +5,59 @@
+        result.push_back(literal);
+      }
+      else
+      {
+        literal += " " + item;
+      }
```
```suggestion
        literal += " " + item;
```

lindsayad commented on framework/src/mfem/functormaterials/MFEMGenericFunctorVectorMaterial.C:
```diff
@@ -5,6 +5,59 @@
+    else if (item.front() == '{')
+    {
+      if (item.back() == '}')
+      {
+        result.push_back(item.substr(1, item.size() - 2));
+      }
```
```suggestion
        result.push_back(item.substr(1, item.size() - 2));
```

lindsayad commented on framework/src/mfem/functormaterials/MFEMGenericFunctorVectorMaterial.C:
```diff
@@ -5,6 +5,59 @@
+    {
+      if (item.back() == '}')
+      {
+        result.push_back(item.substr(1, item.size() - 2));
```
Why is it -2 here instead of -1 ?

lindsayad replied:
But earlier in the function you use `item.size() - 1` to remove a closing bracket

lindsayad replied:
Lol I even looked up the c++ reference for substr and didn't register this
> The second argument is not the final index but the length

lindsayad commented on framework/src/mfem/functormaterials/MFEMGenericFunctorVectorMaterial.C:
```diff
@@ -5,6 +5,59 @@
+      }
+    }
+    else
+    {
+      result.push_back(item);
+    }
+  }
+  if (in_literal)
+  {
+    mooseError("No closing curly brace for vector value in "
+               "MFEMGenericFunctorVectorMaterial prop_values: '{" +
+               literal + "'");
+  }
```
```suggestion
      result.push_back(item);
  }
  if (in_literal)
    mooseError("No closing curly brace for vector value in "
               "MFEMGenericFunctorVectorMaterial prop_values: '{" +
               literal + "'");
```

lindsayad commented on framework/src/mfem/utils/CoefficientManager.C:
```diff
@@ -102,29 +103,67 @@ CoefficientManager::declareMatrixProperty(const std::string & name,
 }
 
+std::shared_ptr<mfem::Coefficient>
+CoefficientManager::getScalarCoefficientPtr(const std::string name)
```
```suggestion
CoefficientManager::getScalarCoefficientPtr(const std::string & name)
```

lindsayad commented on unit/src/MFEMMaterialTest.C:
```diff
@@ -371,10 +217,12 @@ TEST_F(MFEMFunctorScalarMaterialTest, MFEMGenericFunctorMaterial_Exception)
+  coef_params.set<std::vector<std::string>>("prop_names") = {
+      "coef1", "coef2", "constant1", "constant2"};
+  coef_params.set<std::vector<MFEMVectorCoefficientName>>("prop_values") = {
+      "func1", "func2", "{0.", "1.", "2.}", "{1.5}"};
```
What happens when only a single value is supplied like with `"{1.5}"`?

lindsayad replied:
Are all three components set to 1.5 or just the x-component?

cmacmackin replied:
`{1.5}` is interpreted as a single-component vector. If it gets used along with vectors with more components then the missing components of this one will be treated as 0. So, in effect, we are just setting the x-component.

cmacmackin commented on framework/include/utils/MooseStringUtils.h:
```diff
@@ -10,6 +10,7 @@
 #pragma once
 
 #include <algorithm>
+#include <sstream>
```
At some point I'd had an issue with a non-unity build. I can't remember exactly why I concluded this was the best place to put the include though.

cmacmackin replied:
OK, I've looked this up in the CI logs. The templated [MooseUtils::tokenizeAndConvert](https://mooseframework.inl.gov/docs/doxygen/moose/namespaceMooseUtils.html#aadb0b7c436697b8d7ded35fb34109bc9) function creates a local `std::stringstream` variable, which requires the `sstream` header to have been included. It seems that previously this template was only invoked in files that had done this, but it's fragile to assume that other people will know to do so.

cmacmackin committed:
Added class description

lindsayad commented:
Looks like we have some conflicts now. Do you want to resolve them @cmacmackin ? If not, I can do it

nmnobre commented:
Closing as completed via #30803.
